src/discretize.jl

```diff
@@ -586,87 +585,172 @@ function SciMLBase.symbolic_discretize(pde_system::PDESystem,
+        pinnrep.loss_functions = PINNLossFunctions(bc_loss_functions, pde_loss_functions,
+                                        full_likelihood_function, additional_loss, 
+                                        datafree_pde_loss_functions,
+                                        datafree_bc_loss_functions)
+    else
+
+        function full_loss_function(θ, p)
+            # the aggregation happens on cpu even if the losses are gpu, probably fine since it's only a few of them
+            pde_losses = [pde_loss_function(θ) for pde_loss_function in pde_loss_functions]
+            bc_losses = [bc_loss_function(θ) for bc_loss_function in bc_loss_functions]
+
+            # this is kind of a hack, and means that whenever the outer function is evaluated the increment goes up, even if it's not being optimized
+            # that's why we prefer the user to maintain the increment in the outer loop callback during optimization
+            ChainRulesCore.@ignore_derivatives if self_increment
+                iteration[1] += 1
             end
-            logscalar(pinnrep.logger, sum_weighted_pde_losses,
-                      "weighted_loss/sum_weighted_pde_losses", iteration[1])
-            logscalar(pinnrep.logger, sum_weighted_bc_losses,
-                      "weighted_loss/sum_weighted_bc_losses", iteration[1])
-            logscalar(pinnrep.logger, full_weighted_loss,
-                      "weighted_loss/full_weighted_loss",
-                      iteration[1])
-            logvector(pinnrep.logger, adaloss.pde_loss_weights,
-                      "adaptive_loss/pde_loss_weights",
-                      iteration[1])
-            logvector(pinnrep.logger, adaloss.bc_loss_weights,
-                      "adaptive_loss/bc_loss_weights",
-                      iteration[1])
-        end end
-
-        return full_weighted_loss
-    end
 
-    pinnrep.loss_functions = PINNLossFunctions(bc_loss_functions, pde_loss_functions,
-                                               full_loss_function, additional_loss,
-                                               datafree_pde_loss_functions,
-                                               datafree_bc_loss_functions)
+            ChainRulesCore.@ignore_derivatives begin reweight_losses_func(θ, pde_losses,
+                                                                        bc_losses) end
+
+            weighted_pde_losses = adaloss.pde_loss_weights .* pde_losses
+            weighted_bc_losses = adaloss.bc_loss_weights .* bc_losses
+
+            sum_weighted_pde_losses = sum(weighted_pde_losses)
+            sum_weighted_bc_losses = sum(weighted_bc_losses)
+            weighted_loss_before_additional = sum_weighted_pde_losses + sum_weighted_bc_losses
+
+            full_weighted_loss = if additional_loss isa Nothing
+                weighted_loss_before_additional
+            else
+                function _additional_loss(phi, θ)
```

Comment:
Is this unchanged?

Reply:
yes

Reply:
Okay then use dispatch so it's not diff'd